lib/lint.ml

```diff
@@ -0,0 +1,310 @@
+      in
+      (pkg, WeakChecksum err)
+    in
+    let check_one_url ~ctx url =
+      let filename = OpamUrl.to_string (OpamFile.URL.url url) in
+      let checksums = OpamFile.URL.checksum url in
+      match checksums with
+      | [] -> [ err ~ctx ~filename "no checksum" ]
+      | _ ->
+          if
+            List.for_all
+              (fun hash ->
+                match OpamHash.kind hash with `MD5 -> true | _ -> false)
+              checksums
+          then [ err ~ctx ~filename "only MD5 as checksum" ]
+          else []
+    in
+    let check_extra_file (basename, hash) =
+      match OpamHash.kind hash with
+      | `MD5 ->
+          let filename = OpamFilename.Base.to_string basename in
+          [ err ~ctx:"extra-files" ~filename "only MD5 as checksum" ]
+      | _ -> []
+    in
+    let extra_src_errs =
+      List.concat
+        (List.map
+           (fun (_, url) -> check_one_url ~ctx:"extra-sources" url)
+           (OpamFile.OPAM.extra_sources opam))
+    and url_errs =
+      Option.default []
+        (Option.map (check_one_url ~ctx:"url") (OpamFile.OPAM.url opam))
+    and extra_file_errs =
+      Option.default []
+        (Option.map
+           (fun efs -> List.concat (List.map check_extra_file efs))
+           (OpamFile.OPAM.extra_files opam))
+    in
+    url_errs @ extra_file_errs @ extra_src_errs
+
+  let check_no_pin_depends ~pkg opam =
+    match OpamFile.OPAM.pin_depends opam with
+    | [] -> []
+    | _ -> [ (pkg, PinDepends) ]
+
+  let check_no_extra_files ~pkg opam =
+    match OpamFile.OPAM.extra_files opam with
+    | None | Some [] -> []
+    | Some _ -> [ (pkg, ExtraFiles) ]
+
+  let with_temp_dir prefix f =
+    let dir = D.create_temp_dir prefix in
+    let res = f dir in
+    D.remove_dir dir;
+    res
+
+  let get_dune_project_version ~pkg url =
+    with_temp_dir "lint-dune-project-version-" @@ fun dir ->
+    let res =
+      OpamProcess.Job.run
+      @@ OpamRepository.pull_tree
+           (OpamPackage.to_string pkg)
+           (OpamFilename.Dir.of_string dir)
+           (OpamFile.URL.checksum url)
+           [ OpamFile.URL.url url ]
+    in
+    match res with
+    | OpamTypes.Not_available (_, msg) -> Error msg
+    | Up_to_date _ | Result _ -> (
+        let dune_project = Filename.concat dir "dune-project" in
+        try
+          let ch = open_in dune_project in
+          try
+            let content = really_input_string ch (in_channel_length ch) in
+            close_in ch;
+            match Sexplib.Sexp.parse content with
+            | Sexplib.Sexp.Done
+                (List [ Atom "lang"; Atom "dune"; Atom version ], _) ->
+                Ok (Some version)
+            | Done _ -> failwith "(lang dune ...) is not the first construct"
+            | Cont _ -> failwith "Failed to parse the dune-project file"
+          with e ->
+            close_in_noerr ch;
+            raise e
+        with
+        | Sys_error _ -> Ok None
+        | e -> raise e)
```

Comment:
I think this could probably be cleaned up using `In_channel.with_open_*` too? E.g. with something like

```suggestion
      let dune_project = Filename.concat dir "dune-project" in
      match In_channel.input_all |> In_channel.with_open_text dune_project |> Sexplib.Sexp.parse with
      | exception Sys_error _ -> Ok None
      | Sexplib.Sexp.Done (List [ Atom "lang"; Atom "dune"; Atom version ], _) ->
        Ok (Some version)
      | Done _ -> Error "(lang dune ...) is not the first construct"
      | Cont _ -> Error "Failed to parse the dune-project file"
```